WIP: Demonstrate remote signing w/ lightning-signer

ksedgwic opened the pull request:
#### Pull Request Checklist

- [ ] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [ ] All changes are Go version 1.12 compliant
- [ ] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [ ] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [ ] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [ ] Any new logging statements use an appropriate subsystem and
  logging level
- [ ] Code has been formatted with `go fmt`
- [ ] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [ ] Running `make check` does not fail any tests
- [ ] Running `go vet` does not report any issues
- [ ] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [ ] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

Roasbeef committed:
lncfg: add new GetBackends which returns active DB backends

In this commit, we modify the existing `GetBackend` method to now be
called `GetBackends`. This new method will populate a new `RemoteDB`
attribute based on if the replicated backend is active or not. As is,
the local backend is used everywhere. An upcoming commit will once again
re-enable the remote backend, in a hybrid manner.

Roasbeef committed:
multi: split database storage into remote and local instances

In this commit, we split the database storage into two classes: remote
and local data. If etcd isn't active, then everything is actually just
local though we use two pointers everywhere. If etcd is active, then
everything but the graph goes into the remote database.

bhandras committed:
channeldb: move makeTestDB out of test to make it available for other tests

This commit moves makeTestDB to db.go and exports it so that we'll be
able to use this function in other unit tests to make them testable with
etcd if needed.

bhandras committed:
lnd: use channeldb.MakeTestDB in nursery store tests

bhandras committed:
sweep: use channeldb.MakeTestDB

bhandras committed:
lnd: utxo nursery test to use channeldb.MakeTestDB

bhandras committed:
kvdb+etcd: change flattened bucket key derivation algorithm

This commit changes the key derivation algo we use to emulate buckets
similar to bbolt. The issue with prefixing keys with either a bucket or
a value prefix is that the cursor couldn't effectively iterate trough
all keys in a bucket, as it skipped the bucket keys.
While there are multiple ways to fix that issue (eg. two pointers,
iterating value keys then bucket keys, etc), the cleanest is to instead
of prefixes in keys we use a postfix indicating whether a key is a
bucket or a value. This also simplifies all operations where we
(recursively) iterate a bucket and is equivalent with the prefixing key
derivation with the addition that bucket and value keys are now
continous.

bhandras committed:
kvdb: make etcd tests use testify require instead of assert

bhandras committed:
kvdb: s/hu/en/g

bhandras committed:
etcd: remove the lock set concept

This commit removes the lock set which was used to only add bucket keys
to the tx predicate while also bumping their mod version.
This was useful to reduce the size of the compare set but wasn't useful
to reduce contention as top level buckets were always in the lock set.

bhandras committed:
kvdb+etcd: assert on bucket/value key when putting value/bucket

This commit extends compatibility with the bbolt kvdb implementation,
which returns ErrIncompatibleValue in case of a bucket/value key
collision. Furthermore the commit also adds an extra precondition to the
transaction when a key doesn't exist. This is needed as we fix reads to
a snapshot revision and other writers may commit the key otherwise.

cfromknecht committed:
lntest/itest: add whitelist entry for block hash fetch

guggero committed:
mod: update to latest btcd version

guggero committed:
make+itest: make itest Windows compatible

guggero committed:
travis: add itest on Windows

This commit adds an integration test that runs on a Windows virtual
machine on Travis. The tests run inside of a "Git Bash" environment
which supports the same command line syntax as a proper bash but doesn't
have all the tooling installed. Some tools also behave differently on
Windows. Therefore we also have to simplify the command to upload the
logs to termbin and remove the upload to file.io on Windows because both
the find and tar command don't work as expected.

guggero committed:
lntest: improve fee calculation in multi-hop test

guggero committed:
channeldb: fix for Windows clock resolution

We use the event timestamp of a forwarding event as its primary storage
key. On systems with a bad clock resolution this can lead to collisions
of the events if some of the timestamps are identical. We fix this
problem by shifting the timestamps on the nanosecond level until only
unique values remain.

cfromknecht committed:
Merge pull request #4411 from bhandras/test_channeldb

kvdb+etcd+tests: change etcd flattened bucket key derivation to make it compatible with bbolt

ksedgwic committed:
Make InitRequest, validate NodeIDs

halseth committed:
mobile: remember walletunlocker.proto

Roasbeef committed:
Merge pull request #4504 from halseth/mobile-wallet-unlocker-split

mobile: remember walletunlocker.proto

Roasbeef committed:
lnd: don't set freelist value when creating channeldb

This value actually isn't read anywhere, since it's no longer used.
Instead, `cfg.Db.Bolt.NoSyncFreeList` is what's evaluated when we go to
open the DB.

Roasbeef committed:
config: unify old and new config values for sync-freelist

In this commit, unify the old and new values for `sync-freelist`, and
also ensure that we don't break behavior for any users that're using the
_old_ value.

To do this, we first rename what was `--db.bolt.no-sync-freelist`, to
`--db.bolt.sync-freelist`. This gets rid of the negation on the config
level, and lets us override that value if the user is specifying the
legacy config option.

In the future, we'll deprecate the old config option, in favor of the
new DB scoped option.

Roasbeef committed:
lnd: log bbolt freelist sync config value on start up

Kixunil committed:
config: Updated deprecation message of noseedbackup

According to the recent discussion `noseedbackup` is not deprecated.
This change clarifies the message about deprecation.
Also fixes a typo.

Closes #4499

cfromknecht committed:
lntest: add canceled adding new channel to whitelist

devrandom commented on remotesigner/interface.go:
```diff
@@ -0,0 +1,73 @@
+	rsp, err := client.Init(ctx, &InitRequest{
+		Chainparams: &ChainParams{NetworkName: "testnet"},
+		Coldstart:   true,
+		HsmSecret:   &Secret{Data: useSeed},
```
this is now `BIP32Seed`, right?

CandleHater committed:
rpc: add missing space to error message

This corrects the output of the chain notifier RPC error. It has been displayed as: "chain notifier RPC *isstill* in the process of starting"

cfromknecht committed:
Merge pull request #4506 from Roasbeef/free-list-sync-config-fix

config: make bbolt freelist settings mutually exclusive

cfromknecht committed:
Merge pull request #4509 from CandleHater/patch-2

rpc: add missing space to error message

cfromknecht committed:
lntest/itest: add WS connection reset to error whitelist

guggero committed:
lnrpc: add macaroon workaround for WebSockets in browsers

For security reasons, browsers are limited in the header fields they can
send when opening a WebSocket connection. Specifically, the macaroon
cannot be sent in the Grpc-Metadata-Macaroon header field as that would
be possible for normal REST requests. Instead we only have the special
field "Sec-Websocket-Protocol" that can be used to transport custom
data. We allow the macaroon to be sent there and transform it into a
proper header field for the target request.

guggero committed:
docs: describe how to use WebSockets with the REST API

We add a new document that shows two examples of how to use the
WebSocket REST API with JavaScript.

guggero committed:
lntest: add WS test case with custom header macaroon

cfromknecht committed:
lntest/itest: add chain notifier shutting down to whitelist

Roasbeef committed:
Merge pull request #4475 from guggero/travis-windows

travis: add integration test on Windows

Roasbeef committed:
Merge pull request #4511 from guggero/browser-websocket-fix

Bugfix: Allow browsers to send macaroon in WebSocket calls

Roasbeef committed:
Merge pull request #4507 from Kixunil/patch-2

Updated deprecation message of noseedbackup

ksedgwic committed:
Changed remotesigner.nodeID to array, improved logging.

cfromknecht committed:
lntest/itest: add failed getting UTXO error to whitelist

Roasbeef committed:
Merge pull request #4348 from Roasbeef/etcd-local-graph-db

multi: create split local/remote database structure

cfromknecht committed:
build/version: bump version to v0.11.0-beta.rc2

ksedgwic committed:
Added remotesigner.RemoteSignerECDH which shadows and validates an embedded PubKeyECDH.

cfromknecht committed:
Merge pull request #4496 from cfromknecht/moar-whitelist

lntest/itest: update error whitelist

Roasbeef committed:
lnrpc/invoicesrpc: move hop hint selection into new method

Roasbeef committed:
lnrpc/invoicesrpc: extend hop hint selection to account for MPP

In this commit, we update the hop hint selection to account for the fact
that with MPP, a single payment may consume multiple channels. As is, if
a user only has two 0.5 BTC channels, and tries to make a 1 BTC channel,
then the current logic won't include any hop hints.

To solve this, we first add all the channels which in isolation can
carry the payment in question. We then do another pass that accumulates
channels until either we reach our hop-hint limit, or the total
bandwidth that we've accumulate is greater than 2x the payment amount.

Roasbeef committed:
lnrpc/invoicesrpc: move hop hint eligibility for chan to new method

Roasbeef committed:
lnrpc/invoicesrpc: only try to add hop hints if we have channels

ksedgwic committed:
Shadowed SignAnnouncement {ChannelUpdate, NodeAnnouncement}.
ChannelAnnouncement not encountered by integration tests.

guggero committed:
rpcserver: don't allow closing restored channels

To avoid the scenario where the user tries to force close a channel too
early that was restored through SCB, we check the channel status in the
RPC server already. If a restored channel cannot connect to its peer and
mark it as local data loss the channel arbitrator would try to publish
the commitment TX it does not have and crash.

guggero committed:
itest: move channel backup tests to own file

guggero committed:
lntest: make sure restored chans can't be closed

Roasbeef committed:
lnd: properly pass peer remote db for graph, local db for rest

In this commit, we fix a mistake in the split for the new `peer`
package/struct when instantiating the config needed. The existing code
had the DB's swapped. In this commit, we fix this to pass the remote DB
for generic channeldb access, and the local DB for channel graph access.

yyforyongyu committed:
trivial: typo fix

cfromknecht committed:
chainntnfs/txnotifier: remove events on teardown

If a concurrent call to cancel is made while the notifier is shutting
down, this would cause a panic (close of closed channel) since the
events are not removed from the notification sets.

Roasbeef committed:
lnrpc/invoicesrpc: refactor out hop hint accumulation into a new function

cfromknecht committed:
Merge pull request #4526 from Roasbeef/db-split-peer-fix

lnd: properly pass peer remote db for graph, local db for rest

Roasbeef committed:
Merge pull request #4529 from cfromknecht/txnotifier-double-cancel

chainntnfs/txnotifier: remove events on teardown

cfromknecht committed:
Merge pull request #4525 from guggero/no-restore-close

rpcserver: no manual close of restored channels

cfromknecht committed:
Merge pull request #4528 from yyforyongyu/fix-typo

trivial: typo fix

Roasbeef committed:
build: update to latest btcwallet version

This version fixes a race condition related to some of the new coin
selection calls.

Fixes #4532.

Roasbeef committed:
lnwallet: ensure ConfirmedBalance holds the coin select mutex

In this commit, we make a new wrapper method around the internal
`WalletController` method to ensure it holds the coin select mutex while
the balance is being computed.

Roasbeef committed:
Merge pull request #4521 from Roasbeef/mpp-hop-hints

lnrpc/invoicesrpc: extend hop hint selection to account for MPP

Roasbeef committed:
Merge pull request #4533 from Roasbeef/btcwallet-update

btcwallet: update btcwallet version, ensure ConfirmedBalance holds the coin select mutex

cfromknecht committed:
build/version: bump to 0.11.0-beta.rc3

Roasbeef committed:
build: update to btcwallet version with dust bug fix

In this commit, we update our btcwallet dep to the latest version. This
version includes a bug fix for dust calculation. Without this bug fix,
users would potentially significantly overpay on fees, as dust was
computed using the desired fee of the transaction rather than the min
relay fee.

Roasbeef committed:
Merge pull request #4534 from Roasbeef/btcwallet-dust-fix

build: update to btcwallet version with dust bug fix

cfromknecht committed:
build/version: bump to v0.11.0-beta.rc4

ksedgwic committed:
Implement remote GetChannelBasepoints (#5)

Implemented remote GetChannelBasepoints

ksedgwic committed:
Add remote ReadyChannel (#6)

* upgraded GetChannelBasepoints debugging

* Added remotesigner.ReadyChannel, improved debugging.

ksedgwic committed:
Pointed to lightning-signer version of btcutil.

ksedgwic committed:
Added remote sign-remote-commitment (#7)

* Added remote sign-remote-commitment

* Replaced option_static_remotekey w/ CommitmentType.
